### ISSUE TITLE
target in `self._update_weights_generic` for experimental feedback

### ISSUE THREAD

**Issue by LeoKonzett** [QPG-MIT/slmsuite#28]
Hey,

I was experimenting with the method `SpotHologram._update_weights()`. 
One thing I noticed is that when using `feedback=experimental`, the target is defined by `self.spot_amp` instead of `self.target`.
When I calculate the norm of `self.spot_amp` using `Hologram._norm()`, this norm isn't unity. 

I wonder if that's an issue as in `Hologram.update_weights_generic()` we divide (that is, compare) the value of `self.spot_amp` with the (normalized) feedback value `self.feedback_corrected`. If one is normalized but the other isn't, the factors by which we scale the weights will all be below (or above, depending on `spot_amp`) unity.
In comparison, when using `feedback=computational`, the target `self.target` is normalized and the scale factors are either below or above unity, as expected.

Apart from this (I don't know if this is the proper place to ask), I tried to use camera feedback and noticed that the non-uniformity of the spots increases (the image diverges). I did some checks and ensured that the weight updates are reasonable (e.g. the weight of a spot that is very dim gets increased by a large amount), and tried out different gains (`power=0.1, power=1.5`, and 0.1 steps in between). Debugging with `take` also yielded good results, the spots are centered in their crop boxes and the feedback values are reasonable. Also the computational feedback works flawlessly. Perhaps you had similar divergence issues and there are some quick fixes that one can try?

Thanks

**Comment by ichristen** [comment]
Hi @LeoKonzett ,

Thanks for this feedback, it's very useful to see what people are doing and what issues are encountered.

- **`_update_weights()` for `spot_amp`:** Yeah, I think this part was changed back and forth and the current documentation (and code) maybe doesn't reflect the intended behavior. I'm adding the documentation tag to this issue. Also, our tests have been using `spot_amp=None`, which I think is an oversight. The intended behavior is:

  - The user feeds something to `spot_amp=` with some random units (not normalized).
  - Despite what the documentation claims (that `spot_amp` is automatically normalized), we want to keep the same units such that the user can easily modify this (still always `numpy`) list if desired and then have the optimization change towards these values.
  - You're right, that `Hologram.update_weights_generic()` is fed an unnormalized target, but the end of this function normalizes the weights ([see here](https://github.com/QPG-MIT/slmsuite/blob/main/slmsuite/holography/algorithms.py#L843)), so I think it still works out.

- **Optimization divergence:** We have noticed in some cases that image holograms (`FeedbackHologram`) diverge at certain points under some `power` values or levels of feedback. We're in the process of improving this. However, we have never found a similar issue for `SpotHologram` in our use-cases with the current code. `SpotHologram` has always been robust in our tests. I suggest the following:

  - Is your `settle_time_s` long enough for your SLM? If the SLM is not allowed enough time to settle, then you'll be measuring the _previous_ result for the phase pattern from the _previous_ iteration. This can lead to unstable optimization. I think the current default is `settle_time_s=0` on main, but maybe we should restore the `settle_time_s=0.3` default which we previously used as good for most SLMs. The user should determine the `settle_time_s` based on their application. Higher precision holograms require longer time (i.e. longer than 1/e). There should be more documentation regarding this, so more reason to add a documentation tag.
  - What camera are you using? It's possible that `.flush()` (which is intended to avoid old frames and thus outdated feedback) is not working as desired for your camera.
  - If the above doesn't work, could you send us the code snippet that generates the spot pattern of interest? I could imagine certain patterns which are "hard" for WGS to generate, and it might be useful for us to debug. I'm adding the bug tag in lieu of this.

Thanks again for this feedback!

**Comment by ichristen** [comment]
Actually, you're right that the lack of `spot_amp` normalization is a bug. Even though the current code fine for the Leonardo feedback (if target is off by a constant `c`, then the result is off by a factor of `c^p` which goes away with normalization), the Nogrette feedback will yield incorrect results. We will fix this by forcing normalization onto `target_amp` locally in `Hologram.update_weights_generic()`. ETA for a patch next week.

**Comment by LeoKonzett** [comment]
Hi @ichristen

thanks for the quick response. I checked both `settle_time_s` and `.flush()` and both are working properly (tested with e.g. blazing and then capturing a frame). 

The pattern I'm generating is similar to the one from the docs webpage.
```
xlist = np.arange(250, min(fs.cam.shape) - 300, 90)  
ylist = np.arange(250, min(fs.cam.shape) - 300, 90)
xgrid, ygrid = np.meshgrid(xlist, ylist)

square = np.vstack((xgrid.ravel(), ygrid.ravel()))  
```
```
from slmsuite.holography.algorithms import SpotHologram
hologram = SpotHologram(shape=(2048, 2048), spot_vectors=square, basis='ij', cameraslm=fs)

# Optimization
comp_kwargs = {"power": 0.7}
hologram.optimize('WGS-Kim', feedback='computational', stat_groups=['computational'], maxiter=30, **comp_kwargs)
```
Upon recording with the camera, we get a 3x3 array as intended. Checking the resulting `img` with `subimages = analysis.take(img, vectors=square, size=40, integrate=False)` also yields good results.

```
# Optimization w. cam feedback
exp_kwargs = {"power": 0.2, "factor": 0.5} # Factor is for Nogrette

hologram.optimize('WGS-Kim', maxiter=5, feedback='experimental_spot', stat_groups=['computational', 'experimental_spot'], **exp_kwargs)
```

**Comment by LeoKonzett** [comment]
Hi @ichristen,

I managed to spot the issue, it was with me using the wrong coordinate convention. When e.g. blazing to a spot `(100, 700)`, I was blazing to `row_idx=100, column_idx=700` instead of `row_idx=700, col_idx=100`. 
Maybe this could be added to the docs (there you blaze to `(600, 600)` so it's not obvious that the convention is Cartesian. 

Thanks!

**Comment by ichristen** [comment]
So the experimental holography is converging now? I'm not quite sure how coordinate convention would mess with convergence as the convergence is dealt with internally with uniform convention.

Here's my TODO list for this issue's thread:

- [ ] Add more clarity about (numpy) `shape` `(y,x)=(h,w)` vs `vector` `(x,y)` formalism in both docs and examples.
- [ ] Normalize `spot_amp` in `Hologram.update_weights_generic()` and update associated documentation.
- [ ] Figure out what do to with `settle_time_s`.
- [ ] I'm also going to add a `Hologram` flag which will tell stats to save the feedback images/etc with the option to export as an .h5 file for user reference and debugging. That way, people can send us the feedback images so we can check if WGS is doing something stupid.

**Comment by LeoKonzett** [comment]
Hi @ichristen,

yes the holography is converging. 
To be more precise: The camera we're using returns a *transposed* frame. I thought the Fourier calibration accounts for that but it's reasonable that it doesn't. My fix is including a `transpose=True` keyword in the `Camera()` (Camera.py) constructor, similar to `flip_ur=True` or `flip_lr`.

Secondly, there might be an issue with `SpotHologram.make_rectangular_array()`. 
The code snippet 
```
slm_shape = (1200, 1920)  # (h, w) convention
array_holo = SpotHologram.make_rectangular_array(slm_shape, 
array_shape=[8, 4], array_pitch=[12, 12], basis='knm', cameraslm=fs)

zoom = array_holo.plot_farfield(source=array_holo.target, title='Initialized Nearfield')
```
doesn't generate a centered array. Implementing the change below fixes this behavior:
```
# Make the grid edges. This accounts for (h, w) -> (x, y) convention.
        x_edge = (np.arange(array_shape[0]) - (array_shape[0] - 1) / 2)
        x_edge = x_edge * array_pitch[0] + array_center[1]  # changed array_center[0]
        y_edge = (np.arange(array_shape[1]) - (array_shape[1] - 1) / 2)
        y_edge = y_edge * array_pitch[1] + array_center[0] # changed array_center[1]
```

**Comment by ichristen** [comment]
Hi @LeoKonzett,

Thank you for catching the `make_rectangular_array` centering issue. Recently, the default centering logic changed and it looks like I didn't test non-square shapes. The issue is actually on line [1942](https://github.com/QPG-MIT/slmsuite/blob/e572c473a2e8bb141ce5ee428d3ce7727890bdb6/slmsuite/holography/algorithms.py#L1942), and is fixed in e572c473a2e8bb141ce5ee428d3ce7727890bdb6. The grid edge logic is correct because array_center in other units is not flipped.

I'm still a bit confused about the transpose issue. Fourier calibration should have caught that because a transpose is just a 90 deg rotation plus a flip (so I think the current arguments of `make_transform` cover transpose). In testing, I would take the eight cases and make sure Fourier calibration worked on each. The exposure of the camera, the choice of pitch, or noise of the camera might influence how well the "two spots removed" parity check is working, and it just so happens that it works better when the camera is transposed (i.e. something is failing, but the default option happens to be the transposed option so it all works out). If you send the image used for Fourier calibration, I can test the eight cases on that image and make sure things are working robustly.

**Comment by LeoKonzett** [comment]
Hi @ichristen,

the routine below works for `flip_up=True` and `rot="90"`.

```
fs_optimize_kwargs = {"method": "WGS-Kim", "maxiter": 30, "feedback":'computational', "stat_groups":['computational']}

fs.fourier_calibrate(array_shape=[20, 20], array_pitch=[12, 8], array_center=None, plot=True, 
                     autofocus=False, autoexposure=False, **fs_optimize_kwargs)
```
I attach the image I use for calibration below. I slightly adapted `analysis.blob_detect()` to do *dynamic* thresholding (see snippet below) so you might have to change the thresholds for the CV2 blob detector to get similar results.
```
# in analysis.blob_detect() - set CV2 blob detector thresholds dynamically

from skimage.feature import peak_local_max  # import this 

peak_coords = peak_local_max(cv2img, min_distance=15, num_peaks=8)
y_idx = peak_coords[:, 0]
x_idx = peak_coords[:, 1]
heights = cv2img[y_idx, x_idx]

# Set parameters of CV2 simple blob detector dynamically
params.minThreshold = np.min(height_peaks) - 10  # 10 is padding but can be changed
logging.info(f"Lower threshold for CV2 blob detector is {params.minThreshold}")
params.maxThreshold = np.sort(height_peaks)[-2] + 10 
logging.info(f"Upper threshold for CV2 blob detector is {params.minThreshold}")
```
This is not optimized for efficiency.

To test the calibration, I blaze with this code snippet
```
vector_blaze = fs.ijcam_to_kxyslm((100, 700))  # (100, 700) is supposed to be (x, y) convention

blaze_phase = toolbox.blaze(grid=fs.slm, vector=vector_blaze)

plot_phase(blaze_phase, title="Blaze at pixel (100, 100)", zoom=False)  # plot_phase func. from https://slmsuite.readthedocs.io
/en/latest/_examples/experimental_holography.html
```
As visible below, the points gets blazed to `(100, 700)`, which is the intended behavior.

![image](https://user-images.githubusercontent.com/56349940/217540312-5bf570ce-eb4f-4632-a688-c5d7304c6832.png)



![img_for_calibration](https://user-images.githubusercontent.com/56349940/217537935-fc010f75-07da-467e-bb49-581ca46d050b.png)

One **question** that I have: 
In the target image, the two spots for `orientation_check` are missing in the bottom right corner. But in the image above the two spots are missing from the *top left* corner. Is this intended?

**Comment by ichristen** [comment]
Hi @LeoKonzett,

Sorry for my late reply.

 - _I slightly adapted analysis.blob_detect() to do dynamic thresholding (see snippet below) so you might have to change the thresholds for the CV2 blob detector to get similar results._
I like the idea of improvements to `.blob_detect()`; that part of the code hasn't seen much attention in the revamp.  Feel free to improve this function and submit a pull request! However, I would prefer to not add dependencies to `main` (such as `skimage`) unless absolutely necessary, as increasing numbers of dependencies increases bloat and risk of associated dependency errors. Actually, I wanted at one point to remove `cv2` (opencv-python) as a dependency before being convinced otherwise. Consider instead looking for similar features in existing dependencies or writing something simple and custom if functionality is desired.

 - _In the target image, the two spots for orientation_check are missing in the bottom right corner. But in the image above the two spots are missing from the top left corner. Is this intended?_
Yes. The spots are deleted from the bottom-right corner in the $k$-space of the SLM, but your camera seems to have a few flips and rotations that causes the empty spots to appear in the top-left. This is part of the reason to have Fourier calibration: so the user doesn't have to care or think about these flips and rotations.
Three more things:

    - We tend to get more reliable results when we overexpose (saturate) the Fourier calibration image to reduce the influence of the 0th order peak on calibration. However, I'm going to poke at your image in a few days to see if I can get it to run more reliably in the general unsaturated case.
    - Your spots look a bit distorted. Have you run [wavefront calibration](https://slmsuite.readthedocs.io/en/latest/_examples/wavefront_calibration.html)?
    - Your image has an interesting checkerboard pattern at the pixel level. Is this intentional?

**Comment by ichristen** [comment]
### `(x, y)` vs `(y, x)`.

The first docs reference you linked is unclear. I updated these docs for clarity (ea312473d12a3e525d9ed2a44d27f2c9b7238af7). This is what's actually happening:

- Any array / image `shape` follows the `numpy` `(h, w)` [convention](https://stackoverflow.com/questions/43272848/what-is-dimension-order-of-numpy-shape-for-image-data). 
- Any vector or position / etc uses `(x, y)` convention.

This is the formalism that `numpy`, `scipy`, `matplotlib`, etc take. The `shape` and indexing of an array is always inverted, but other functions such as `numpy.meshgrid(x, y)` (default), `scipy.odr.Data(x, y)`, or `matplotlib.pyplot.scatter(x, y)` use standard cartesian `(x, y)` that is more familiar to users. Yes, it's not ideal, but this is the formalism generally adopted by the community.

### Gridpoint rounding.

The spots are rounding to the nearest `"knm"` gridpoint. This is expected behavior. The computational $k$-space cannot have infinite resolution, as this would cost infinite memory (and take a while to Fourier transform). This is why you might want to use padding (`calculate_padded_shape()`) to effectively enhance the resolution of the SLM farfield. See the first tip in the constructor of hologram to learn more about padding. The Fourier calibration pads by default.

Another important note: the plotted image is the target nearfield of the camera and the target farfield of the SLM. It isn't the nearfield of the SLM.

Hope this helps!

**Comment by ichristen** [comment]
Hi Leo,

The TODO list from Feb 6th was merged in with dev (#35), so I'm going to close this issue. Thanks again for raising this!
